x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,338 @@
+// but this is a safety check, since extra invalidation is only a performance impact, not a
+// correctness impact.
+func (oa *OIDCAuthenticator) invalidateAccessToken(conn driver.Connection) {
+	oa.mu.Lock()
+	defer oa.mu.Unlock()
+	tokenGenID := conn.OIDCTokenGenID()
+	// If the connection used in a Reauth is a new connection it will not have a correct tokenGenID,
+	// it will instead be set to 0. In the absence of information, the only safe thing to do is to
+	// invalidate the cached accessToken.
+	if tokenGenID == 0 || tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		conn.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context, cfg *Config) error {
+	oa.invalidateAccessToken(cfg.Connection)
+	return oa.Auth(ctx, cfg)
+}
+
+// Auth authenticates the connection.
+func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	conn := cfg.Connection
+
+	oa.mu.Lock()
+	cachedAccessToken := oa.accessToken
+	oa.mu.Unlock()
+
+	if cachedAccessToken != "" {
+		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
+			userName:    oa.userName,
+			accessToken: cachedAccessToken,
+		})
+		if err == nil {
+			return nil
+		}
+		// this seems like it could be incorrect since we could be inavlidating an access token that
+		// has already been replaced by a different auth attempt, but the TokenGenID will prevernt
+		// that from happening.
+		oa.invalidateAccessToken(conn)
+		time.Sleep(invalidateSleepTimeout)
+	}
+
+	if oa.OIDCHumanCallback != nil {
+		return oa.doAuthHuman(ctx, cfg, oa.OIDCHumanCallback)
+	}
+
+	// Handle user provided or automatic provider machine callback.
+	var machineCallback OIDCCallback
+	if oa.OIDCMachineCallback != nil {
+		machineCallback = oa.OIDCMachineCallback
+	} else {
+		machineCallback, err = oa.providerCallback()
+		if err != nil {
+			return fmt.Errorf("error getting built-in OIDC provider: %w", err)
+		}
+	}
+
+	if machineCallback != nil {
+		return oa.doAuthMachine(ctx, cfg, machineCallback)
+	}
+	return newAuthError("no OIDC callback provided", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthHuman(_ context.Context, _ *Config, _ OIDCCallback) error {
+	// TODO GODRIVER-3246: Implement OIDC human flow
+	// Println is for linter
+	fmt.Println("OIDC human flow not implemented yet", oa.idpInfo)
+	return newAuthError("OIDC human flow not implemented yet", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthMachine(ctx context.Context, cfg *Config, machineCallback OIDCCallback) error {
+	subCtx, cancel := context.WithTimeout(ctx, machineCallbackTimeout)
+	accessToken, err := oa.getAccessToken(subCtx,
```

Comment:
The line in the OIDC spec that this timeout comes from is ambiguous with respect to the Go Driver:
> If CSOT is not applied, then the driver MUST use 1 minute as the timeout.

All Go Driver blocking APIs support timeouts, so from that perspective, CSOT is always applied. However, the Go Driver also includes the concept of CSOT being "enabled" or not, which was required to retrofit the CSOT behavior into the existing driver that already supported timeouts (for example, see usage of `csot.IsTimeoutContext`). From that perspective, we should only apply the 1-minute timeout when CSOT is not "enabled".

As implemented, the code always applies the 1-minute timeout, overriding any longer timeouts. That seems unlikely to break anything because waiting 1 minute for an access token will probably cause other problems (e.g. the default `connectTimeout` is 30 seconds).

Overall, I think this code is safe, but is worth a comment describing the interpretation of the OIDC spec we're using here.

E.g. comment:
> The CSOT spec says to apply a 1-minute timeout if "CSOT is not applied". That's ambiguous for the v1.x Go Driver because it could mean either "no timeout provided" or "CSOT not enabled". Always use a maximum timeout duration of 1 minute, allowing us to ignore the ambiguity. Contexts with a shorter timeout are unaffected.

Reply:
This is exactly the sort of context I am missing, thank you!